staging/src/k8s.io/code-generator/cmd/validation-gen/validators/required.go

```diff
@@ -98,15 +134,39 @@ const (
 )
 
 var (
-	optionalValidator = types.Name{Package: libValidationPkg, Name: "Optional"}
+	optionalValueValidator   = types.Name{Package: libValidationPkg, Name: "OptionalValue"}
+	optionalPointerValidator = types.Name{Package: libValidationPkg, Name: "OptionalPointer"}
+	optionalSliceValidator   = types.Name{Package: libValidationPkg, Name: "OptionalSlice"}
+	optionalMapValidator     = types.Name{Package: libValidationPkg, Name: "OptionalMap"}
 )
 
 func (optionalDeclarativeValidator) ExtractValidations(t *types.Type, comments []string) (Validations, error) {
 	_, optional := gengo.ExtractCommentTags("+", comments)[optionalTagName]
 	if !optional {
 		return Validations{}, nil
 	}
-	return Validations{Functions: []FunctionGen{Function(optionalTagName, ShortCircuit|NonError, optionalValidator)}}, nil
+	// Most validators don't care whether the value they are validating was
+	// originally defined as a value-type or a pointer-type in the API.  This
+	// one does.  Since Go doesn't do partial specialization of templates, we
+	// do manual dispatch here.
+	for t.Kind == types.Alias {
+		t = t.Underlying
+	}
+	switch t.Kind {
+	case types.Slice:
+		return Validations{Functions: []FunctionGen{Function(optionalTagName, ShortCircuit|NonError, optionalSliceValidator)}}, nil
+	case types.Map:
+		return Validations{Functions: []FunctionGen{Function(optionalTagName, ShortCircuit|NonError, optionalMapValidator)}}, nil
+	case types.Pointer:
+		return Validations{Functions: []FunctionGen{Function(optionalTagName, ShortCircuit|NonError, optionalPointerValidator)}}, nil
+	case types.Struct:
```

Comment:
I think this same logic applies to Required and Forbidden, too?

Reply:
For required/Forbidden, if the type is struct, shouldn't we also validate it through (requiredValue/ForbiddenValue)? It is different from the "+optional"

Reply:
For a non-pointer struct, what do these tags mean?

Optional: If ALL of the fields (transitively) are zero-value, then don't do any further validation.  If ANY one field is set (transitively) then do all validation.  That's pretty weird, and I will argue that the right way to express it is on the individual fields.  But it is useful documentation.

Required: If ALL of the fields (transitively) are zero-value, then don't do any further validation.  If ANY one field is set (transitively) then this tag is satisfied; do all validation.  I will again argue that the right way to express it is on the individual fields.  But it is useful documentation.  E.g. all metav1.ObjectMeta SHOULD be marked as required, not optional.

Forbidden: If ALL of the fields (transitively) are zero-value, then this tag is satisfied.  If ANY one field is set (transitively) then it is an error.  We don't even have this tag today, and I can only see utility in it with cross-field validation.  That said, it is slightly more sensical than the others.  I think we should probably disallow the use of it for non-pointer structs and see if a use-case emerges.

Reply:
That's also my understanding on all these 3. So, for required and forbidden, non-pointer struct field validation should use `SomethingValue()` instead of omitting it like for +optional.

Reply:
Updated the code for the non-pointer struct

Reply:
Repeat for required and `return Validations{}, fmt.Errorf("non-pointer structs cannot use the %q tag")` with a good comment about "if you hit this and need think you need it to work, we need to understand why"